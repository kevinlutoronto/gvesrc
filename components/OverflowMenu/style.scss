@import "../../styles/base";

.container {
  position: relative;
}

.action {
  height: 35px;
  cursor: pointer;
  display: flex;
  align-items: center;
}

.revealable {
  position: absolute;
  z-index: 1000;
  display: none;
  top: 0;
  left: 0;
  &.openLeft {
    left: auto;
  }
}

.revealed {
  display: block;
}

.revealableTopWrapper {
  box-sizing: border-box;
  position: absolute;
  top: -3px;
  left: -4px;
  overflow: hidden;
  padding: 3px 4px 0 4px;
  height: 33px;
  z-index: 10;
  cursor: pointer;
}

.revealableTop {
  box-sizing: border-box;
  height: 35px;
  background-color: $cWhite;
  box-shadow: 0 2px 4px 0 rgba(0, 0, 0, 0.2);
  display: flex;
  justify-content: center;
  align-items: center;
}

.overflow {
  @extend %icon--overflow;
}

.labelWithText {
  color: #252729;
  font-size: 13px;
  font-weight: bold;
  line-height: 20px;
  text-align: center;
  padding-left: 3px;
  padding-right: 3px;
  white-space: nowrap;
}

.revealableList {
  cursor: pointer;
  box-sizing: border-box;
  background-color: $cWhite;
  position: relative;
  z-index: 1;
  padding: 10px 0;
  margin-top: 30px;
  box-shadow: 0 0px 4px 0 rgba(0, 0, 0, 0.2);
  min-width: 200px;
  &.openLeft {
    left: -100%;
    margin-right: -40px;
  }
}

.revealableListItem {
  @extend %type--body;
  box-sizing: border-box;
  background-color: $cWhite;
  list-style: none;
  padding: 5px 10px;
  cursor: pointer;
  &:hover {
    background-color: $cGray600;
  }
}
