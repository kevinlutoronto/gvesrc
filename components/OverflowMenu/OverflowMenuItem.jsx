// @flow
import React from 'react';
import style from './style.scss';

type Props = {
  onClick: string => void,
  optionKey: string,
  text: React$Node,
};

const OverflowMenuItem = ({ onClick, optionKey, text }: Props) => {
  return (
    <li className={style.revealableListItem} onClick={() => onClick(optionKey)}>
      {text}
    </li>
  );
};

OverflowMenuItem.displayName = 'Plasma@OverflowMenuItem';

export default OverflowMenuItem;
