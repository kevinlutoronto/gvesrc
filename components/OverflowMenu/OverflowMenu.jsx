// @flow
import cx from 'classnames';
import { map } from 'lodash';
import React from 'react';
import { getDataAttrs } from '../../dataUtils';
import type { Data } from '../../types';
import OverflowMenuItem from './OverflowMenuItem';
import style from './style.scss';

const direction = { RIGHT: 'right', LEFT: 'left' };

type Props = {|
  disabled?: boolean,
  options: Array<{| key: string, text: React$Node |}>,
  onClick: string => void,
  openDirection?: $Values<typeof direction>,
  label?: string,
  data?: Data,
|};

type State = {|
  revealed: boolean,
|};

class OverflowMenu extends React.Component<Props, State> {
  static defaultProps = {
    options: [],
    openDirection: direction.RIGHT,
  };

  state = {
    revealed: false,
  };

  handleMouseEnter = (): void => {
    if (!this.props.disabled) {
      this.setState({ revealed: true });
    }
  };

  handleMouseLeave = (): void => {
    this.setState({ revealed: false });
  };

  handleClick = (optionKey: string) => {
    this.setState({ revealed: false });
    this.props.onClick(optionKey);
  };

  renderLabel = (): React$Node => {
    const { label } = this.props;
    if (label) {
      return <div className={style.labelWithText}>{label} &#9662;</div>;
    }

    return <div className={style.overflow} />;
  };

  render() {
    const revealableStyle = cx(style.revealable, {
      [style.revealed]: this.state.revealed,
    });
    const revealableListStyle = cx(style.revealableList, {
      [style.openLeft]: this.props.openDirection && this.props.openDirection === direction.LEFT,
    });
    return (
      <div
        {...getDataAttrs(this.props.data)}
        onMouseEnter={this.handleMouseEnter}
        onMouseLeave={this.handleMouseLeave}
        className={style.container}
      >
        <div className={style.action}>{this.renderLabel()}</div>
        <div className={revealableStyle}>
          <div className={style.revealableTopWrapper}>
            <div className={style.revealableTop}>{this.renderLabel()}</div>
          </div>
          <ol className={revealableListStyle}>
            {map(this.props.options, option => {
              return (
                <OverflowMenuItem
                  key={option.key}
                  optionKey={option.key}
                  text={option.text}
                  onClick={this.handleClick}
                />
              );
            })}
          </ol>
        </div>
      </div>
    );
  }
}

OverflowMenu.displayName = 'Plasma@OverflowMenu';

export default OverflowMenu;
