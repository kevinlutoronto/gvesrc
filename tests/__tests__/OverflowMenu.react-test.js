/* eslint-disable */

import React from 'react';
import OverflowMenu from '../../components/OverflowMenu/OverflowMenu';
import renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';

describe('OverflowMenu Component', () => {
  test('render', () => {
    const component = renderer.create(<OverflowMenu />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('render with label', () => {
    const component = renderer.create(<OverflowMenu label="Hello World!" />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
