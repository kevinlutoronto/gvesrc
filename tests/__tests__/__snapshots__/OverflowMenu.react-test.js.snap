// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OverflowMenu Component render 1`] = `
<div
  className="container"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <div
    className="action"
  >
    <div
      className="overflow"
    />
  </div>
  <div
    className="revealable"
  >
    <div
      className="revealableTopWrapper"
    >
      <div
        className="revealableTop"
      >
        <div
          className="overflow"
        />
      </div>
    </div>
    <ol
      className="revealableList"
    />
  </div>
</div>
`;

exports[`OverflowMenu Component render with label 1`] = `
<div
  className="container"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <div
    className="action"
  >
    <div
      className="labelWithText"
    >
      Hello World!
       ▾
    </div>
  </div>
  <div
    className="revealable"
  >
    <div
      className="revealableTopWrapper"
    >
      <div
        className="revealableTop"
      >
        <div
          className="labelWithText"
        >
          Hello World!
           ▾
        </div>
      </div>
    </div>
    <ol
      className="revealableList"
    />
  </div>
</div>
`;
